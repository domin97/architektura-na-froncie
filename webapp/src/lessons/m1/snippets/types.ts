export type Currency = 'USD' | 'GBP' | 'EUR' | 'PLN'

export interface Employee {
  id: string
  firstName: string
  lastName: string
  salary: number
  title: string
}
