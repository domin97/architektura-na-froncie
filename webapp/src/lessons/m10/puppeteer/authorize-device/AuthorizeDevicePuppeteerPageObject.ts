import { Page } from 'puppeteer'

const getByTestId = (testId: string) => `[data-testid="${testId}"]`

export const selectors = {
  linkAuthorizeDevice: getByTestId('link-authorize-device'),
  btnAuthorizeDevice: getByTestId('btn-authorize-device'),
  chooseAddDeviceButton: getByTestId('btn-choose-add-device'),
  chooseAllowOnceButton: getByTestId('btn-choose-allow-once'),
  chooseLogoutButton: getByTestId('btn-choose-logout'),
  addDeviceNameInput: getByTestId('input-add-device-name'),
  addDeviceNameSubmitButton: getByTestId('btn-add-device-name-submit'),
  deviceAddConfirmationCheckbox: getByTestId('checkbox-add-device-confirmation'),
  tokenCancelButton: getByTestId('btn-token-cancel'),
  tokenSubmitButton: getByTestId('btn-token-submit'),
  tokenResetButton: getByTestId('btn-token-reset'),
  addDevicePasswordInput: getByTestId('input-add-device-password'),
  allowOncePasswordInput: getByTestId('input-allow-once-password'),
  closeButton: getByTestId('btn-close'),
}

export class AuthorizeDevicePuppeteerPageObject {
  constructor(
    private page: Page
  ){}

  async navigateToPage(){
    return this.page.goto('http://localhost:3010', { waitUntil: 'networkidle2' })
  }

  async startAuthorization(){
    await this.page.click(selectors.linkAuthorizeDevice)
    return this.page.click(selectors.btnAuthorizeDevice)
  }

  async clickChooseAddDeviceButton(){
    return this.page.click(selectors.chooseAddDeviceButton)
  }

  async clickChooseAllowOnceButton(){
    return this.page.click(selectors.chooseAllowOnceButton)
  }

  async clickLogoutButton(){
    return this.page.click(selectors.chooseLogoutButton)
  }

  private async clearInput(inputSelector: string){
    await this.page.evaluate((selector) => {
      const example = document.querySelector(selector);
      (example as HTMLInputElement).value = '';
    }, inputSelector);
  }

  async confirmAddDeviceName(deviceName?: string){
    if (deviceName){
      await this.clearInput(selectors.addDeviceNameInput)
      await this.page.click(selectors.addDeviceNameInput)
      await this.page.keyboard.type(deviceName)
    }
    await this.page.click(selectors.deviceAddConfirmationCheckbox)
    return this.page.click(selectors.addDeviceNameSubmitButton)
  }

  async setDevicePasswordPassword(password: string){
    await this.page.click(selectors.addDevicePasswordInput)
    return this.page.keyboard.type(password)
  }

  async submitAddDeviceToken(password: string){
    await this.setDevicePasswordPassword(password)
    return this.clickTokenSubmitButton()
  }

  async setAllowOncePassword(password: string){
    await this.page.click(selectors.allowOncePasswordInput)
    return this.page.keyboard.type(password)
  }

  async submitAllowOnceToken(password: string){
    await this.setAllowOncePassword(password)
    return this.clickTokenSubmitButton()
  }

  async clickTokenCancelButton(){
    return this.page.click(selectors.tokenCancelButton)
  }

  async clickTokenSubmitButton(){
    return this.page.click(selectors.tokenSubmitButton)
  }

  async clickTokenResetButton(){
    return this.page.click(selectors.tokenResetButton)
  }

  async clickCloseButton(){
    return this.page.click(selectors.closeButton)
  }

  async shouldHaveTextDisplayed(text: string){
    return this.page.waitForFunction(
      (text: string) => document.querySelector('body')!.innerText.includes(text),
      {},
      text
    );
  }

  async shouldHaveTextDisappeared(text: string){
    return this.page.waitForFunction(
      (text: string) => ! document.querySelector('body')!.innerText.includes(text),
      {},
      text
    );
  }
}
