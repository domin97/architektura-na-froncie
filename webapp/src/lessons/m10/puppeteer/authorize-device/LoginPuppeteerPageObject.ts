import { Page } from 'puppeteer'
import { shouldHaveTextDisappeared } from '../utils'

const getByTestId = (testId: string) => `[data-testid="${testId}"]`

export const selectors = {
  inputLoginUsername: getByTestId('input-login-username'),
  inputLoginPassword: getByTestId('input-login-password'),
  btnLogin: getByTestId('btn-login'),
}

export class LoginPuppeteerPageObject {
  constructor(
    private page: Page
  ){}

  async navigateToPage(){
    return this.page.goto('http://localhost:3010', { waitUntil: 'networkidle2' })
  }

  async fillCredentials(username: string, password: string){
    await this.page.click(selectors.inputLoginUsername)
    await this.page.keyboard.type(username)

    await this.page.click(selectors.inputLoginPassword)
    await this.page.keyboard.type(password)
  }

  async clickLoginButton(){
    return this.page.click(selectors.btnLogin)
  }

  async shouldHaveTextDisappeared(text: string){
    return shouldHaveTextDisappeared(this.page, text)
  }
}
