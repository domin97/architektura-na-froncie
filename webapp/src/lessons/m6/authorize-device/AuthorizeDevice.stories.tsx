import React, { useState } from 'react';

import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { AuthorizeDeviceProcessPrimitive } from './hooks/AuthorizeDeviceProcessPrimitive'
import { AuthorizeDeviceProcessUnion } from './hooks/AuthorizeDeviceProcessUnion'
import { AuthorizeDeviceProcessReduxWithStore } from './redux/AuthorizeDeviceProcessRedux'
import { AuthorizeDeviceProcessXState } from './xstate/AuthorizeDeviceProcessXState'
import { tokenMockHandlers } from 'api/mock/token.mock';

import { Panel } from 'ui/layout';
import { Button } from 'ui/atoms';
import { Modal } from 'ui/molecules';


export default {
  title: 'Lessons/M6 State Machines/Authorize Device',
  parameters: {
    msw: tokenMockHandlers
  },
} as Meta;

const notify = action('notify')

type AuthorizeDeviceProcessComponent = React.FC<{
  onSuccess: () => void
  onLogout: () => void
}>

// story factory (HOC)
const generateInline = (Component: AuthorizeDeviceProcessComponent) => {
  return () => {
    const [status, setStatus] = useState<"CLEAN" | "RUNNING" | "SUCCESS" | "LOGGED_OUT">("CLEAN")
    const start = () => { setStatus("RUNNING") }

    const successFn = () => {
      setStatus("SUCCESS")
      notify('succeeded')
    }

    const logoutFn = () => {
      setStatus("LOGGED_OUT")
      notify('logged out')
    }

    return <Panel>
      { (status === "SUCCESS") && <div>authorization succeeded</div> }
      { (status === "LOGGED_OUT") && <div>logged out</div> }
      { (status !== "RUNNING") && <Button
          data-testid="btn-authorize-device"
          variant="PRIMARY"
          onClick={start}
        >autoryzuj urzÄ…dzenie</Button>
      }
      { (status === "RUNNING") && <Component
        onSuccess={successFn}
        onLogout={logoutFn}
      /> }
    </Panel>
  }
}

export const _AuthorizeDeviceProcessPrimitive = generateInline(AuthorizeDeviceProcessPrimitive)
export const _AuthorizeDeviceProcessUnion = generateInline(AuthorizeDeviceProcessUnion)
export const _AuthorizeDeviceProcessRedux = generateInline(AuthorizeDeviceProcessReduxWithStore)
export const _AuthorizeDeviceProcessXState = generateInline(AuthorizeDeviceProcessXState)

export const InModal = () => {
  const [isRunning, setRunning] = useState(false)
  const start = () => { setRunning(true) }

  const successFn = () => {
    setRunning(false)
    notify('succeeded')
  }

  const logoutFn = () => {
    setRunning(false)
    notify('logged out')
  }

  return !isRunning
    ? <Button
        variant="PRIMARY"
        onClick={start}
      >add device</Button>
    : <Modal
        isOpen={isRunning}
        onRequestClose={logoutFn}
      >
        <AuthorizeDeviceProcessUnion
          onSuccess={successFn}
          onLogout={logoutFn}
        />
      </Modal>
}
