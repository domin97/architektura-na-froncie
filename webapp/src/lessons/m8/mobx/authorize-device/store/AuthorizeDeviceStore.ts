import { makeAutoObservable, makeObservable, observable, action, computed } from 'mobx'
import { StateMember } from 'lessons/m6/state-members'
import { AuthorizeDeviceState, AuthorizeDeviceProcess } from 'lessons/m6/authorize-device/types'

import { getTokenInstruction, sendTokenCode } from 'api/token'

export class AuthorizeDeviceStore implements AuthorizeDeviceProcess {
  constructor(){
    makeObservable(this, {
      _state: observable,
      setState: action,
      state: computed,
    })
  }

  public _state: AuthorizeDeviceState = {
    type: "CHOOSE_METHOD"
  }

  setState(newState: AuthorizeDeviceState){
    this._state = newState
  }

  get state(){
    return this._state
  }

  // analogicznie jak w generycznej funkcji "assertState"
  // ale dodatkowo dodajemy constraint TType na dostępne (istniejące) typy
  assertState<TType extends AuthorizeDeviceState['type']>(
    state: { type: string },
    expectedType: TType
  ): asserts state is StateMember<TType> {
    if (state.type !== expectedType){
      throw new Error(`Invalid state ${state.type} (expected: ${expectedType})`)
    }
  }

  cancelChoice = async () => {
    this.setState({ type: "CHOOSE_METHOD" })
  }

  chooseAllowOnce = async () => {
    this.assertState(this._state, "CHOOSE_METHOD")

    this.setState({ type: "LOADING" })
    const tokenInstruction = await getTokenInstruction()
    this.setState({
      type: "ALLOW_ONCE_TOKEN",
      ...tokenInstruction,
      error: false
    })
  }

  submitAllowOnce = async (password: string) => {
    this.assertState(this._state, "ALLOW_ONCE_TOKEN")

    // UWAGA! ustawienie stanu na "loading" może wyczyścić aktualny stan
    // (bo zostałby podmieniony - dlatego zmienne lokalne)
    const instruction = this._state.instruction
    const tokenId = this._state.tokenId
    // zobaczmym, że `assertState` powyżej
    // (1) rzuca errory, jeśli metoda jest w błędnym stanie oraz
    // (2) doprecyzowuje (w compile-time) typ aktualnego stanu, które w tym wypadku jest BŁĘDNE :(
    this.setState({ type: "LOADING" })
    // (2) this._state // w compile-time jest "ALLOW_ONCE_TOKEN" co jesty nieprawdą
    // `this._state` oraz `this.setState` w runtime odnoszą się do jednej rzeczy, ale kompilator widzi je jako osobne - i traci type flow
    // jaka jest lepsza opcja?
    // moglibyśmy dodać `this.assertState(this._state, "LOADING")` albo usunąć assertState na samym początku - ale wtedy łamiemy (1) :(
    // inna opcja to wprowadzić osobne stany LOADING obsługujące różne przejścia między stanami - podobnie jak robi xstate (większy porządek, ale i więcej stanów/więcej roboty)
    try {
      await sendTokenCode({ tokenId, tokenCode: password })
      this.setState({
        type: "ALLOW_ONCE_SUCCESS"
      })
    } catch (e: unknown) {
      this.setState({
        type: "ALLOW_ONCE_TOKEN",
        instruction,
        tokenId,
        error: true,
      })
    }
  }

  chooseAddDevice = async () => {
    this.assertState(this._state, "CHOOSE_METHOD")
    this.setState({
      type: "ADD_DEVICE_FORM",
    })
  }

  submitDeviceName = async (currentDeviceName: string) => {
    this.assertState(this._state, "ADD_DEVICE_FORM")

    this.setState({ type: "LOADING" })
    const tokenInstruction = await getTokenInstruction()
    this.setState({
      type: "ADD_DEVICE_TOKEN",
      deviceName: currentDeviceName,
      ...tokenInstruction,
      error: false,
    })
  }

  resetToken = async () => {
    this.assertState(this._state, "ADD_DEVICE_TOKEN")

    // watch out! setting state to loading might reset the deviceName
    const deviceName = this._state.deviceName
    this.setState({ type: "LOADING" })
    const tokenInstruction = await getTokenInstruction()
    this.setState({
      type: "ADD_DEVICE_TOKEN",
      deviceName,
      ...tokenInstruction,
      error: false,
    })
  }

  submitAddDevice = async (password: string) => {
    this.assertState(this._state, "ADD_DEVICE_TOKEN")

    // watch out! setting state to loading might reset the deviceName
    const deviceName = this._state.deviceName
    const instruction = this._state.instruction
    const tokenId = this._state.tokenId
    this.setState({ type: "LOADING" })
    try {
      await sendTokenCode({ tokenId, tokenCode: password })
      this.setState({
        type: "ADD_DEVICE_CONFIRMATION",
        deviceName,
      })
    } catch (e: unknown) {
      this.setState({
        type: "ADD_DEVICE_TOKEN",
        deviceName,
        instruction,
        tokenId,
        error: true,
      })
    }
  }

  confirmDeviceAdded = async () => {
    this.assertState(this._state, "ADD_DEVICE_CONFIRMATION")
    this.setState({
      type: "ADD_DEVICE_SUCCESS",
    })
  }
}
