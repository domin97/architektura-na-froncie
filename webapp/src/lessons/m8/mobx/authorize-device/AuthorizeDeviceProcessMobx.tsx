import React, { useEffect } from 'react';
import { reaction } from "mobx"
import { observer } from "mobx-react-lite"

import {
  AuthorizeDeviceChooseMethodView,
  AuthorizeDeviceAllowOnceTokenView,
  AuthorizeDeviceAddDeviceFormView,
  AuthorizeDeviceAddDeviceTokenView,
  AuthorizeDeviceAddDeviceConfirmationView,
} from 'ui/authorize-device/views'
import { AuthorizeDeviceStore } from './store/AuthorizeDeviceStore';
import { Loader } from 'ui/atoms';

interface AuthorizeDeviceProcessMobxProps {
  onSuccess: () => void
  onLogout: () => void
  store: AuthorizeDeviceStore
}

export const AuthorizeDeviceProcessMobx = observer<AuthorizeDeviceProcessMobxProps>(
  props => {
    const { store, onSuccess, onLogout } = props

    // 🔥 reaction jest "odmianą" autoruna, która uruchamiana jest tylko dla (ZMIENIONEGO) kawałka stanu
    // i na tym kawałku jest potem uruchamiany side effect
    useEffect(() => { // 🔥 reaktywność WIDOKU
      return reaction( // 🔥 reaktywność STANU
        () => store.state,
        (state) => {
          if (state.type === 'ALLOW_ONCE_SUCCESS'){
            onSuccess()
          }
        }
      )
    }, [store, onSuccess])

    switch(store.state.type){
      case "LOADING":
        return <Loader />

      case "CHOOSE_METHOD":
        return <AuthorizeDeviceChooseMethodView
          onAddDeviceToTrusted={store.chooseAddDevice}
          onAllowDeviceOnce={store.chooseAllowOnce}
          onLogout={onLogout}
        />

      /**
       * 🔥 UWAGA!
       * Lepiej dla szeroko pojętej wydajności byłoby przekazywać w dół do komponentów
       * NIE KONKRETNE WARTOŚCI (np. store.state.instruction) a referencję na całęgo store'a (patrz: slajd "MobX & React - dobre praktyki")
       * My tutaj robimy wyjątek tylko "szkoleniowo" - chcemy pozostać przy 1 implementacji widoków (które są state-agnostic), i które możemy reużyć w wielu implementacjach (primitive obsession, state union, redux, xstate, mobx...)
       * Innymi słowy, naszym celem jest mieć 5 rozwiązań dla procesu + 5 komponentów-widoków (reużywalnych, dla czytelności) ZAMIAST 5 rozwiązań + 25 komponentów-widoków
       *
       * Czyli w projekcie (nie-szkoleniowym) dalibyśmy:
       * <AuthorizeDeviceAllowOnceTokenView store={store} /> - jeśli przekazujemy store'a przez propsy
       * lub
       * <AuthorizeDeviceAllowOnceTokenView /> - jeśli store ciągniemy przez Context API
       */
       case "ALLOW_ONCE_TOKEN":
        return <AuthorizeDeviceAllowOnceTokenView
          onSubmit={store.submitAllowOnce}
          onCancel={store.cancelChoice}
          instruction={store.state.instruction}
          error={store.state.error}
        />

      case "ALLOW_ONCE_SUCCESS":
        return null

      case "ADD_DEVICE_FORM":
        return <AuthorizeDeviceAddDeviceFormView
          onSubmit={store.submitDeviceName}
        />

      case "ADD_DEVICE_TOKEN":
        return <AuthorizeDeviceAddDeviceTokenView
          deviceName={store.state.deviceName}
          instruction={store.state.instruction}
          onSubmit={store.submitAddDevice}
          onReset={store.resetToken}
          onCancel={store.cancelChoice}
          error={store.state.error}
        />

      case "ADD_DEVICE_CONFIRMATION":
        return <AuthorizeDeviceAddDeviceConfirmationView
          deviceName={store.state.deviceName}
          onClose={onSuccess}
        />

      case "ADD_DEVICE_SUCCESS":
        return null

      case "LOGGED_OUT":
        return null

      default:
        const leftover: never = store.state
        return null
      }
  })
