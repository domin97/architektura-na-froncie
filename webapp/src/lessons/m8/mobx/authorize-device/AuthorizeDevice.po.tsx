import { render } from "@testing-library/react";

import { unsafe_cast } from "lib/lang";

import { AuthorizeDevicePO } from "lessons/m7/authorize-device/AuthorizeDevice.po";
import { AuthorizeDeviceProcessMobx as AuthorizeDeviceProcess } from "./AuthorizeDeviceProcessMobx";
import { AuthorizeDeviceStore } from "./store/AuthorizeDeviceStore";

/**
 * UWAGA! üî•
 *
 * Dlaczego dziedziczenie Page Objecta?
 * - szkoleniowo tak wygodniej
 * Poni≈ºszy Page Object renderuje komponent nieco inaczej ni≈º oryginalny Page Object (z modu≈Çu o testowaniu integracyjnym).
 * Ten tutaj potrzebuje store MobXowy, musi go r√≥wnie≈º stworzyƒá.
 * Mogliby≈õmy skopiowaƒá ca≈Ço≈õƒá... ale r√≥≈ºni≈Çaby siƒô jedynie metoda render, jak poni≈ºej.
 * W PRAWDZIWEJ APLIKACJI nie mieliby≈õmy potrzeby takiego dziedziczenia, bo implementacja wystarczy jedna.
 */
export class AuthorizeDeviceMobXPO extends AuthorizeDevicePO {
  static render() {
    const onSuccess = jest.fn()
    const onLogout = jest.fn()
    const store = new AuthorizeDeviceStore()

    const { container } = render(<AuthorizeDeviceProcess
      onSuccess={onSuccess}
      onLogout={onLogout}
      store={store}
    />)

    return new AuthorizeDevicePO(unsafe_cast.ElementToHTMLElement(container), onSuccess, onLogout);
  }
}
