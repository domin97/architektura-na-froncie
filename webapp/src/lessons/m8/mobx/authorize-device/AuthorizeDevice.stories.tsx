import React, { useState } from 'react';

import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { AuthorizeDeviceProcessMobx as AuthorizeDeviceProcess } from './AuthorizeDeviceProcessMobx'
import { AuthorizeDeviceStore } from './store/AuthorizeDeviceStore';
import { tokenMockHandlers } from 'api/mock/token.mock';

import { Panel } from 'ui/layout';
import { Button } from 'ui/atoms'


export default {
  title: 'Lessons/M8 Reactive/MobX',
  parameters: {
    msw: tokenMockHandlers
  },
  argTypes: {}
} as Meta;

const notify = action('notify')

// no boolean obsession
export const AuthorizeDevice = () => {
  const [status, setStatus] = useState<"CLEAN" | "RUNNING" | "SUCCESS" | "LOGGED_OUT">("CLEAN")
  const start = () => { setStatus("RUNNING") }

  const successFn = () => {
    setStatus("SUCCESS")
    notify('succeeded')
  }

  const logoutFn = () => {
    setStatus("LOGGED_OUT")
    notify('logged out')
  }

  return <Panel>
    { (status === "SUCCESS") && <div>authorization succeeded</div> }
    { (status === "LOGGED_OUT") && <div>logged out</div> }
    {!(status === "RUNNING")
    ? <Button
        variant="PRIMARY"
        onClick={start}
      >autoryzuj urzÄ…dzenie</Button>
    : <AuthorizeDeviceProcess
        onSuccess={successFn}
        onLogout={logoutFn}
        store={new AuthorizeDeviceStore()}
      />}
  </Panel>
}
