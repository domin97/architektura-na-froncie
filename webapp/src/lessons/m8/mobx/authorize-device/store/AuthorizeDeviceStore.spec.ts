import { autorun } from "mobx";

import { AuthorizeDeviceProcess } from "lessons/m6/authorize-device/types";
import { AuthorizeDeviceStore } from "./AuthorizeDeviceStore";

// import { getTokenInstruction, sendTokenCode } from '../../../../../api/token'
jest.mock('../../../../../api/token', () => ({
  getTokenInstruction: async () => ({
    instruction: "Wpisz hasÅ‚o SMS",
    tokenId: "1111-2222-3333-4444"
  }),
  sendTokenCode: async (params: { tokenCode: string }) => {
    if (params.tokenCode.length !== 4){
      throw new Error(`Invalid confirmation token!`)
    }
  },
}))

describe("AuthorizeDeviceStore", () => {
  const configure = () => {
    const store: AuthorizeDeviceProcess = new AuthorizeDeviceStore();
    const spy = jest.fn();
    autorun(() => spy(store.state));
    return { store, spy }
  }

  describe('allow device once', () => {

    it("should notify state changes when successfully allowing device once", async () => {
      // given
      const { store, spy } = configure()

      // when
      await store.chooseAllowOnce();
      await store.submitAllowOnce("1234");

      // then
      expect(spy.mock.calls).toMatchSnapshot();
    });

    it("should notify state changes when failing to allow device once", async () => {
      // given
      const { store, spy } = configure()

      // when
      await store.chooseAllowOnce();
      await store.submitAllowOnce("too long");

      // then
      expect(spy.mock.calls).toMatchSnapshot();
    });
  })

  describe('add device', () => {
    it("should notify state changes when successfully adding device", async () => {
      // given
      const { store, spy } = configure()

      // when
      await store.chooseAddDevice();
      await store.submitDeviceName("my computer");
      await store.resetToken();
      await store.submitAddDevice("1234");
      await store.confirmDeviceAdded();

      // then
      expect(spy.mock.calls).toMatchSnapshot();
    });

    it("should notify state changes when failing to add device", async () => {
      // given
      const { store, spy } = configure()

      // when
      await store.chooseAddDevice();
      await store.submitDeviceName("my computer");
      await store.submitAddDevice("too long");

      // then
      expect(spy.mock.calls).toMatchSnapshot();
    })
  });
});
