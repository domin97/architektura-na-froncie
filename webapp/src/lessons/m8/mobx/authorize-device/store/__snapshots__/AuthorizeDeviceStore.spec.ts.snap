// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorizeDeviceStore add device should notify state changes when failing to add device 1`] = `
Array [
  Array [
    Object {
      "type": "CHOOSE_METHOD",
    },
  ],
  Array [
    Object {
      "type": "ADD_DEVICE_FORM",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "deviceName": "my computer",
      "error": false,
      "instruction": "Wpisz hasło SMS",
      "tokenId": "1111-2222-3333-4444",
      "type": "ADD_DEVICE_TOKEN",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "deviceName": "my computer",
      "error": true,
      "instruction": "Wpisz hasło SMS",
      "tokenId": "1111-2222-3333-4444",
      "type": "ADD_DEVICE_TOKEN",
    },
  ],
]
`;

exports[`AuthorizeDeviceStore add device should notify state changes when successfully adding device 1`] = `
Array [
  Array [
    Object {
      "type": "CHOOSE_METHOD",
    },
  ],
  Array [
    Object {
      "type": "ADD_DEVICE_FORM",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "deviceName": "my computer",
      "error": false,
      "instruction": "Wpisz hasło SMS",
      "tokenId": "1111-2222-3333-4444",
      "type": "ADD_DEVICE_TOKEN",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "deviceName": "my computer",
      "error": false,
      "instruction": "Wpisz hasło SMS",
      "tokenId": "1111-2222-3333-4444",
      "type": "ADD_DEVICE_TOKEN",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "deviceName": "my computer",
      "type": "ADD_DEVICE_CONFIRMATION",
    },
  ],
  Array [
    Object {
      "type": "ADD_DEVICE_SUCCESS",
    },
  ],
]
`;

exports[`AuthorizeDeviceStore allow device once should notify state changes when failing to allow device once 1`] = `
Array [
  Array [
    Object {
      "type": "CHOOSE_METHOD",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "error": false,
      "instruction": "Wpisz hasło SMS",
      "tokenId": "1111-2222-3333-4444",
      "type": "ALLOW_ONCE_TOKEN",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "error": true,
      "instruction": "Wpisz hasło SMS",
      "tokenId": "1111-2222-3333-4444",
      "type": "ALLOW_ONCE_TOKEN",
    },
  ],
]
`;

exports[`AuthorizeDeviceStore allow device once should notify state changes when successfully allowing device once 1`] = `
Array [
  Array [
    Object {
      "type": "CHOOSE_METHOD",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "error": false,
      "instruction": "Wpisz hasło SMS",
      "tokenId": "1111-2222-3333-4444",
      "type": "ALLOW_ONCE_TOKEN",
    },
  ],
  Array [
    Object {
      "type": "LOADING",
    },
  ],
  Array [
    Object {
      "type": "ALLOW_ONCE_SUCCESS",
    },
  ],
]
`;
